locallibrary/catalog/models.py

```diff
@@ -1,3 +1,84 @@
 from django.db import models
 
-# Create your models here.
+# Used to generate URLs by reversing the URL patterns
+from django.urls import reverse
+import uuid  # Required for unique book instances
+
+
+class Genre(models.Model):
+    """Model representing a book genre."""
+    name = models.CharField(
+        max_length=200, help_text='Enter a book genre (e.g.Science Fiction)')
+
+    def __str__(self):
+        """String for representing the Model object."""
+        return self.name
+
+
+class Book(models.Model):
+    """Model representing a book (but not a specific copy of a book)."""
+    title = models.CharField(max_length=200)
+    author = models.ForeignKey('Author', on_delete=models.SET_NULL, null=True)
+    summary = models.TextField(
+        max_length=1000, help_text='Enter a brief description of the book')
+    isbn = models.CharField('ISBN', max_length=13, unique=True,
+                            help_text='13 Character <ahref="https://www.isbn-international.org/content/what-isbn'
+                                      '">ISBN number</a>')
+    genre = models.ManyToManyField(
+        Genre, help_text='Select a genre for this book')
+
+    def __str__(self):
+        """String for representing the Model object."""
+        return self.title
+
+    def get_absolute_url(self):
+        """Returns the url to access a detail record for this book."""
+        return reverse('book-detail', args=[str(self.id)])
+
+
+class BookInstance(models.Model):
+    """Model representing a specific copy of a book (i.e. that can be borrowed from the library)."""
+    id = models.UUIDField(primary_key=True, default=uuid.uuid4, help_text='Unique ID for this particular bookacross '
+                                                                          'whole library')
+    book = models.ForeignKey('Book', on_delete=models.RESTRICT)
```

Comment:
Tìm hiểu các giá trị của option `on_delete` e nhé

Reply:
Các giá trị của on_delete: 
- CASCADE:  xóa tất cả các bản ghi tham chiếu đến bản ghi bị xóa.
- SET_NULL: Đặt giá trị các trường là null
- SET_DEFAULT: Đặt giá trị của các trường là mặc định
- PROTECT: Không cho phép xoá
- RESTRICT: Không cho phép xoá bằng cách raise lên RestrictedError. Khác với PROTECT, Việc xoá sẽ được chấp nhận nếu object được xoá thông qua CASCADE 
- DO_NOTHING: Không làm gì cả

Reply:
Tìm hiểu thêm về 3 giá trị `CASCADE, PROTECT, RESTRICT` nè, có ví dụ cụ thể càng tốt em nhé

Reply:
- CASCADE: Là giá trị mặc định. Nó cho phép Django xóa tất cả các đối tượng tham chiếu đối tượng bị xóa.
Ví dụ
`
class Author(models.Model):
    name = models.CharField(max_length = 255)

class Book(models.Model):
    title = models.CharField(max_length = 255)
    author = models.ForeignKey(Author,
                               on_delete=models.CASCADE)

author_1 = Author.objects.create(name=“author_abc”)
book = Book.objects.create(title=“bookabc”, author=author_1)

author_1.delete()
`
-> Sẽ xoá cả book có book.author = author_1

- PROTECT : Không cho phép xoá
Ví dụ:
`
author = models.ForeignKey(Author,
                           on_delete=models.PROTECT)

# Run queries
author_1 = Author.objects.create(name=“author_abc”)
book = Book.objects.create(title=“bookabc”, author=author_1)

author_1.delete()
# Raises Protected Error
# First need to delete all post where post.author=author_1
# then Django will allow to delete author_1
`
-> Raises Protected Error. Cần phải xoá hết book mà có book.author = author_1 xong thì Django mới cho xoá author_1

- RESTRICT : Không cho phép xoá bằng cách raise lên RestrictedError. Khác với PROTECT, Việc xoá sẽ được chấp nhận nếu object được xoá thông qua CASCADE
Ví dụ:
`
class Artist(models.Model):
    name = models.CharField(max_length=10)

class Album(models.Model):
    artist = models.ForeignKey(Artist, on_delete=models.CASCADE)

class Song(models.Model):
    artist = models.ForeignKey(Artist, on_delete=models.CASCADE)
    album = models.ForeignKey(Album, on_delete=models.RESTRICT)
`
`
>>> artist_one = Artist.objects.create(name='artist one')
>>> artist_two = Artist.objects.create(name='artist two')
>>> album_one = Album.objects.create(artist=artist_one)
>>> album_two = Album.objects.create(artist=artist_two)
>>> song_one = Song.objects.create(artist=artist_one, album=album_one)
>>> song_two = Song.objects.create(artist=artist_one, album=album_two)
>>> album_one.delete()
# Raises RestrictedError.  (bởi vì album_one được tham chiếu bởi song_one, mà song_one.album có giá trị on_delete là restrict)
>>> artist_two.delete() (artist_two được tham chiếu bởi album_two, mà album_two được tham chiếu bởi song_two, mà song_two.album có giá trị on_delete là restric)
# Raises RestrictedError.
>>> artist_one.delete() (Có thể xoá)
`

Reply:
- PROTECT : Không cho phép xoá
Ví dụ:
```
author = models.ForeignKey(Author,
                           on_delete=models.PROTECT)

author_1 = Author.objects.create(name=“author_abc”)
book = Book.objects.create(title=“bookabc”, author=author_1)

author_1.delete()
```
-> Raises Protected Error. Cần phải xoá hết book mà có book.author = author_1 xong thì Django mới cho xoá author_1

Reply:
- RESTRICT : Không cho phép xoá bằng cách raise lên RestrictedError. Khác với PROTECT, Việc xoá sẽ được chấp nhận nếu object được xoá thông qua CASCADE
Ví dụ:
```
class Artist(models.Model):
    name = models.CharField(max_length=10)

class Album(models.Model):
    artist = models.ForeignKey(Artist, on_delete=models.CASCADE)

class Song(models.Model):
    artist = models.ForeignKey(Artist, on_delete=models.CASCADE)
    album = models.ForeignKey(Album, on_delete=models.RESTRICT)


>>> artist_one = Artist.objects.create(name='artist one')
>>> artist_two = Artist.objects.create(name='artist two')
>>> album_one = Album.objects.create(artist=artist_one)
>>> album_two = Album.objects.create(artist=artist_two)
>>> song_one = Song.objects.create(artist=artist_one, album=album_one)
>>> song_two = Song.objects.create(artist=artist_one, album=album_two)
>>> album_one.delete()
# Raises RestrictedError.  (bởi vì album_one được tham chiếu bởi song_one, mà song_one.album có giá trị on_delete là restrict)
>>> artist_two.delete() (artist_two được tham chiếu bởi album_two, mà album_two được tham chiếu bởi song_two, mà song_two.album có giá trị on_delete là restric)
# Raises RestrictedError.
>>> artist_one.delete() (Có thể xoá)
```